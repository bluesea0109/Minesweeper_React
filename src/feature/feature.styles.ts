import {makeStyles} from '@mui/styles';

export const useFeatureStyles = makeStyles({
  cell: {
    minWidth: 10,
    minHeight: 10,
    border: 'none',
    borderRadius: 2,
    backgroundColor: "#bfbfbf",
    cursor: 'grab',
  },
  text: {
    fontWeight: 'bold',
    margin: 0,
    textAlign: 'center',
  },
  row: {
    width: '100%',
    gap: 2,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
});
