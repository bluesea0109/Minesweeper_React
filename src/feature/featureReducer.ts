import { createSlice } from "@reduxjs/toolkit";

const onFetchMap = (payload: any): string[] => {
  const rowList = payload.split("map:")[1].split("\n");
  return rowList.filter((item: string[]) => !!item.length);
}

interface FeatureState {
  map: string[];
  mineSign: string[];
  message: string;
}

const initialState: FeatureState = {
  map: [],
  mineSign: [],
  message: "",
};

const featureStore = createSlice({
  name: "feature",
  initialState,
  reducers: {
    initRound(state) {
      state.mineSign = [];
      state.message = "";
    },
    restartRound(state, action) {
      state.mineSign = [];
      state.message = "";
    },
    updateMineSign(state, action) {
      state.mineSign = [...action.payload];
    },
    newRound(state, action) {
    },
    getMap(state) {},
    setMap(state, action) {
      state.map = onFetchMap(action.payload);
    },
    updateMessage(state, action) {
      state.message = action.payload;
    },
  },
});

export const {
  initRound,
  getMap,
  setMap,
  newRound,
  updateMessage,
  updateMineSign,
  restartRound,
} = featureStore.actions;

export const featureReducer = featureStore.reducer;
