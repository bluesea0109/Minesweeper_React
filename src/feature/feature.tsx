import { SocketEnd } from "../socketConnection/socketConfig";
import { useFeatureStyles } from "./feature.styles";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../store";
import { updateMineSign } from "./featureReducer";

interface Props {
  mapData: string[];
}

export const Feature = ({ mapData }: Props) => {
  const dispatch = useDispatch();
  const classes = useFeatureStyles();
  const featureState = useSelector((state: RootState) => state.feature);
  const mineSign = featureState.mineSign;

  /** Cell mouse left click event function */
  const onCellClick = (y: number, x: number, e: any) => {
    SocketEnd.socket.send(`open ${x} ${y}`);
  };

  /** Cell mouse right click event function */
  const onCellRightClick = (y: number, x: number, e: any) => {
    e.preventDefault();
    if (featureState.message === "You lose") {
      return;
    } else if (mineSign.find((element: any) => element === `cell-${y}-${x}`)) {
      dispatch(
        updateMineSign(
          mineSign.filter((item: any) => !`cell-${y}-${x}`.includes(item)),
        ),
      );
    } else {
      dispatch(updateMineSign([...mineSign, `cell-${y}-${x}`]));
    }
  };

  const renderMap = (items: any) => {
    const level = items.length;
    const height = `calc((100vh - 300px) / ${level})`;

    return items.map((item: any, rowIndex: number) => {
      const squares = item.split("");
      const arrLength = squares.length;
      const row = squares.map((square: any, columnIndex: number) => {
        const key = `cell-${rowIndex}-${columnIndex}`;
        const testId = `cell-${rowIndex}-${columnIndex}`;
        if (square !== "â–¡") {
          return (
            <div
              onClick={(e: any) => onCellClick(rowIndex, columnIndex, e)}
              key={key}
              onContextMenu={(e: any) => e.preventDefault()}
              className={classes.cell}
              style={{
                height: height,
                width: height,
                backgroundColor: "#e6e6e6",
              }}
              data-testid={testId}
            >
              {square && square !== "0" && (
                <p
                  className={classes.text}
                  style={{
                    lineHeight: height,
                    fontSize: arrLength !== 100 || arrLength !== 500 ? 18 : 10,
                  }}
                >
                  {square}
                </p>
              )}
            </div>
          );
        }

        return (
          <div
            onClick={(e: any) =>
              !mineSign.find((element) => element === key) &&
              onCellClick(rowIndex, columnIndex, e)
            }
            key={key}
            onContextMenu={(e: any) =>
              onCellRightClick(rowIndex, columnIndex, e)
            }
            className={classes.cell}
            style={{
              height: height,
              width: height,
            }}
            data-testid={testId}
          >
            {mineSign.find((element) => element === key) && (
              <p
                style={{
                  lineHeight: height,
                  margin: 0,
                  fontSize: arrLength !== 100 || arrLength !== 500 ? 18 : 10,
                }}
              >
                ðŸ’£
              </p>
            )}
          </div>
        );
      });

      return (
        <div className={classes.row} key={`cell-row-${rowIndex}`}>
          {row}
        </div>
      );
    });
  };

  return <>{renderMap(mapData)}</>;
};
