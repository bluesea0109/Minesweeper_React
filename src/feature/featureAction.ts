import {
  take, put, call, apply, fork,
} from 'redux-saga/effects';
import {eventChannel} from 'redux-saga';
import {SocketEnd} from '../socketConnection/socketConfig';
import {setMap, updateMessage} from './featureReducer';

function createSocketChannel(socket: any) {
  return eventChannel((emit) => {
    const handleOnMessage = (event: any) => {
      emit(event.data);
    };

    const errorHandler = (errorEvent: any) => {
      emit(new Error(errorEvent?.reason || 'UKNOWN'));
    };

    socket.addEventListener('message', handleOnMessage);
    socket.addEventListener('error', errorHandler);

    const unsubscribe = () => {
      socket.removeEventListener('message', handleOnMessage);
    };

    return unsubscribe;
  });
}

function* getMap(socket: any) {
  yield apply(socket, socket.send, ['map']);
}

export function* ReceivingData(action: any) {
  yield apply(SocketEnd.socket, SocketEnd.socket.send, [action.payload]);
}

export function* watchChannel(): any {
  const socket: any = yield call(SocketEnd.createConnection);
  const socketChannel = yield call(createSocketChannel, socket);

  while (true) {
    try {
      const data = yield take(socketChannel);
      if (data.includes('map:')) {
        yield put(setMap(data));
      }
      if (data.includes('new:')) {
        yield fork(getMap, socket);
      }
      if (data.includes('open:')) {
        yield put(updateMessage(data.split('open: ')[1]));
        yield fork(getMap, socket);
      }
    } catch (err) {
      alert(err);
      console.error('socket error:', err);
      socketChannel.close();
    }
  }
}

