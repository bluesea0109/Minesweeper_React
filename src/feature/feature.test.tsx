import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { Feature } from "./feature";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { cleanup } from "@testing-library/react";

const sagaMiddleware = createSagaMiddleware();
const mockStore = configureStore([sagaMiddleware]);

const initialState = {
  feature: {
    map: [],
    mineSign: [],
    message: ""
  }
};

describe("<Feature />", () => {
  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });
  it(">>>FEATURE--- Snapshot", () => {
    const store = mockStore(initialState);
    const mapData: string[] = [];
    const tree = renderer.create(
        <Provider store={store}>
          <Feature mapData={mapData} />
        </Provider>
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("cell rendering correctly", () => {
    const store = mockStore(initialState);
    const mapData: string[] = [
      "□□□□□□□□□□",
      "□□□□□□□□□□",
      "□□□□□□□□□□",
      "□□□□□□□□□□",
      "□□□□□□□□□□",
      "□□□□□□□□□□",
      "□□□□□□□□□□",
      "□□□□□□□□□□",
      "□□□□□□□□□□",
      "□□□□□□□□□□",
    ];
    const wrapper = mount(      
      <Provider store={store}>
        <Feature mapData={mapData} />
      </Provider>
    );
    const element = wrapper.find("cell-1-1");
    expect(element).toBeTruthy();
  });
});
