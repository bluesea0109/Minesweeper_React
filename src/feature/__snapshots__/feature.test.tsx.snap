// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feature /> >>>FEATURE--- Snapshot 1`] = `null`;
